codegen/src/sbt-test/generator/e2e/src/test/scala/proto/test/OptionalFieldsSpec.scala

```diff
@@ -0,0 +1,31 @@
+package proto.test
+
+import com.twitter.util.Future
+import com.soundcloud.twinagle.ServerBuilder
+
+import org.specs2.mutable.Specification
+
+class OptionalFieldsSpec extends Specification {
+  // if this compiles, we're good
+  "Proto files with optional scalar values are generated correctly" in {
+    val svc = new OptionalFieldsTestService {
+      override def doSomething(req: IHaveOptionalFields): Future[IHaveOptionalFields] = {
+        iTakeAnOptionInt(req.maybeInt)
+        iTakeAnOptionString(req.maybeString)
```

Comment:
this would also pass if `maybeInt` were of type `Some[Int]` or `None`
I think it might be better to assert the fields' types directly (without even needing a service I guess?)

Reply:
I added same-type constraints to these methods, so they can _only_ be called by a value of the exact type we expect.

Reply:
my FP-scala-foo is a bit rusty. What is the difference to writing `private def iTakeAnOptionInt(_: Option[Int[): Unit = ()`?

Reply:
That's what I originally had, but as Jordan pointed out, if the generated code was of type `Some[Int]` or `None[Int]` it would still pass. Using the same-type constraint ensures that the function can really _only_ be called with an expression that has the exact type `Option[T]`.

I'm not sure how much we _really_ need to be concerned about this though, so I'm happy to revert it if you 'd prefer.